pandas-stubs/io/common.pyi

```diff
@@ -1,65 +1,165 @@
+    StringIO,
+    TextIOBase,
+)
+from pathlib import Path
+import tarfile
 from typing import (
     IO,
-    Any,
     AnyStr,
-    Mapping,
+    Generic,
+    Literal,
+    TypeVar,
+    overload,
 )
 import zipfile
 
-from pandas._typing import FilePathOrBuffer
+from pandas._typing import (
+    BaseBuffer,
+    CompressionDict,
+    CompressionOptions,
+    FilePath,
+    ReadBuffer,
+    StorageOptions,
+    WriteBuffer,
+)
+
+_BaseBufferT = TypeVar("_BaseBufferT", bound=BaseBuffer)
+
+class IOArgs:
+    filepath_or_buffer: str | BaseBuffer
+    encoding: str
+    mode: str
+    compression: CompressionDict
+    should_close: bool
+    def __init__(
+        self, filepath_or_buffer, encoding, mode, compression, should_close
+    ) -> None: ...
 
-lzma = ...
+class IOHandles(Generic[AnyStr]):
+    handle: IO[AnyStr]
+    compression: CompressionDict
+    created_handles: list[IO[bytes] | IO[str]]
+    is_wrapped: bool
+    def close(self) -> None: ...
+    def __enter__(self) -> IOHandles[AnyStr]: ...
+    def __exit__(self, *args: object) -> None: ...
+    def __init__(self, handle, compression, created_handles, is_wrapped) -> None: ...
 
-def is_url(url) -> bool: ...
-def validate_header_arg(header) -> None: ...
+def is_url(url: object) -> bool: ...
+def validate_header_arg(header: object) -> None: ...
+@overload
 def stringify_path(
-    filepath_or_buffer: FilePathOrBuffer[AnyStr],
-) -> FilePathOrBuffer[AnyStr]: ...
-def is_s3_url(url) -> bool: ...
-def is_gcs_url(url) -> bool: ...
-def urlopen(*args, **kwargs) -> IO: ...
-def get_filepath_or_buffer(
-    filepath_or_buffer: FilePathOrBuffer,
-    encoding: str | None = ...,
-    compression: str | None = ...,
-    mode: str | None = ...,
-): ...
+    filepath_or_buffer: FilePath, convert_file_like: bool = ...
+) -> str: ...
+@overload
+def stringify_path(
+    filepath_or_buffer: _BaseBufferT, convert_file_like: bool = ...
+) -> _BaseBufferT: ...
+def urlopen(*args, **kwargs): ...
+def is_fsspec_url(url: FilePath | BaseBuffer) -> bool: ...
 def file_path_to_url(path: str) -> str: ...
 def get_compression_method(
-    compression: str | Mapping[str, str] | None
-) -> tuple[str | None, dict[str, str]]: ...
+    compression: CompressionOptions,
+) -> tuple[str | None, CompressionDict]: ...
 def infer_compression(
-    filepath_or_buffer: FilePathOrBuffer, compression: str | None
+    filepath_or_buffer: FilePath | BaseBuffer, compression: str | None
 ) -> str | None: ...
+def check_parent_directory(path: Path | str) -> None: ...
+@overload
 def get_handle(
-    path_or_buf,
+    path_or_buf: FilePath | BaseBuffer,
     mode: str,
-    encoding=...,
-    compression: str | Mapping[str, Any] | None = ...,
+    *,
+    encoding: str | None = ...,
+    compression: CompressionOptions = ...,
+    memory_map: bool = ...,
+    is_text: Literal[False],
+    errors: str | None = ...,
+    storage_options: StorageOptions = ...,
+) -> IOHandles[bytes]: ...
+@overload
+def get_handle(
```

Comment:
Technically, only classes/functions/... listed on this page are considered to be public https://pandas.pydata.org/docs/reference/index.html Since there are many classes/functions that should be public but are not listed there, people seem to assume that classes/functions that do not start with `_` are by common conventions public.

As far as I know, nothing in io/common is meant to be public. Adding it here might suggest that it is public. That is obviously a pandas issue but it would be great to discuss some guidelines of what should/shouldn't be in the stubs.

Reply:
If that is the goal then it is probably a lot more doable.  The stubs now look like they were generated using an earlier version of stubgen.  They have a lot of wrong things in them, including entire files that have no corresponding code in pandas.

Reply:
> Technically, only classes/functions/... listed on this page are considered to be public https://pandas.pydata.org/docs/reference/index.html Since there are many classes/functions that should be public but are not listed there, people seem to assume that classes/functions that do not start with `_` are by common conventions public.
> 
> As far as I know, nothing in io/common is meant to be public. Adding it here might suggest that it is public. That is obviously a pandas issue but it would be great to discuss some guidelines of what should/shouldn't be in the stubs.

We could have that discussion outside this issue.  Do you want to start an issue or a discussion on this topic?  I have some thoughts on it, but would rather discuss it in a more focused area on that topic.

Reply:
> If that is the goal then it is probably a lot more doable. The stubs now look like they were generated using an earlier version of stubgen. They have a lot of wrong things in them, including entire files that have no corresponding code in pandas.

Yes, this is a historical artifact. These stubs were generated by Microsoft, with stubgen, and possibly on pandas 1.1 or 1.2.  I used them heavily when they were only shipped with VS Code, and kept doing PR's there to make my team's code pass things, and then we had discussions in our monthly pandas dev meetings about how to move forward, given there was another effort for stubs that had testing (which is where the tests here came from).  Net result are these stubs, which we all knew would take a lot of work to get right, but provided a good starting point.  Didn't want to wait to make them "perfect".